tendermint/src/abci/data.rs

```diff
@@ -1,64 +1,78 @@
 ///
 /// Transactions are opaque binary blobs which are validated according to
 /// application-specific rules.
-#[derive(Clone, Debug, Eq, PartialEq, Default)]
+#[derive(Clone, Debug, Eq, PartialEq, Default, Serialize, Deserialize)]
+#[serde(try_from = "Base64", into = "Base64")]
 pub struct Data(Vec<u8>);
 
-impl Data {
-    /// Borrow the data as bytes
-    pub fn as_bytes(&self) -> &[u8] {
-        self.0.as_ref()
-    }
-}
+impl TryFrom<Vec<u8>> for Data {
+    type Error = Error;
 
-impl AsRef<[u8]> for Data {
-    fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        // Todo: Are there any restrictions on the incoming data?
+        Ok(Self(value))
     }
 }
 
-impl Display for Data {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        for byte in &self.0 {
-            write!(f, "{:02X}", byte)?;
-        }
-        Ok(())
+impl From<Data> for Vec<u8> {
+    fn from(value: Data) -> Self {
+        value.0
     }
 }
 
-impl FromStr for Data {
-    type Err = Error;
+impl TryFrom<Base64> for Data {
+    type Error = Error;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        // Accept either upper or lower case hex
-        let bytes = hex::decode_upper(s)
-            .or_else(|_| hex::decode(s))
-            .map_err(|_| Kind::Parse)?;
+    fn try_from(value: Base64) -> Result<Self, Self::Error> {
+        value.0.try_into()
+    }
+}
 
-        Ok(Data(bytes))
+impl From<Data> for Base64 {
+    fn from(value: Data) -> Self {
+        Self(value.into())
     }
 }
 
-impl<'de> Deserialize<'de> for Data {
-    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
-        use serde::de::Error;
-        String::deserialize(deserializer)?
-            .parse()
-            .map_err(D::Error::custom)
+impl AsRef<[u8]> for Data {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
     }
 }
 
-impl Serialize for Data {
-    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
-        self.to_string().serialize(serializer)
+#[cfg(test)]
+mod tests {
+    use crate::abci::Data;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn test_deserialization() {
+        let json = "\"ChYKFGNvbm5lY3Rpb25fb3Blbl9pbml0\"";
+        let mydata: Data = serde_json::from_str(json).unwrap();
+        assert_eq!(
+            mydata.0,
+            vec![
+                // By chance this is a protobuf struct.
+                10, // Field 1 is a String
+                22, // Field 1 length is 22
+                10, // Sub-field 1 is String
+                20, // Sub-field 1 length is 20
+                99, 111, 110, 110, 101, 99, 116, 105, 111, 110, 95, 111, 112, 101, 110, 95, 105,
+                110, 105, 116 // "connection_open_init"
+            ]
+        );
+    }
+
+    #[test]
+    fn test_serialization() {
+        let mydata: Data = Data::try_from(vec![1, 2, 3, 4]).unwrap();
```

Comment:
```suggestion
        let mydata = Data::try_from(vec![1, 2, 3, 4]).unwrap();
```

And if you do away with the `TryFrom<Vec<u8>>` implementation above, then it could become:

```rust
        let mydata = Data::from(vec![1, 2, 3, 4]);
```

Reply:
done